### PR TITLE
multiple Lin pots reading

### PR DESCRIPTION
**Pull request by WilliamLim781**
started the refactoring for lin pots removed the polling function, created DMA variables, modified Displacement functions to accept DMA variables, and enabled multiple analog inputs.

### CODE REVIEW

**Comment by WilliamLim781** [comment]
ran some test for the lin pot functions for inches and mm and got these results

for the inches it got these readings

lin pot fully extended 
fully extended

lin pot fully compressed


fully compressed

now testing travel in mm i got 

fully extended
fully extended(mm)


fully compressed

fully compressed(mm)

lin pot has a travel distance of 3 inches or 76.2 mm. These values aren't exactly there but they are close enough to the target value. Some reasons for the deviation could be due to poor tolerances of a potentiometer. there is easily at least a 1-ohm rating when the potentiometer is supposed to read 0 ohms.  The second potential reason would be the noise of the signals when testing jumpers and on a breadboard are not the cleanest of signals. 

also note that the potentiometers I used to get these values were not very stable physically. I was struggling to keep them on the breadboard so that would contribute to why some of the values fluctuate. Also, note that there is a 500ms delay to make these values easy for me to read. This means in some of the pictures you can literally see the value either slowly increase or decrease. That is not an error on the adc but rather just me slowly turning the potentiometer

**Comment by WilliamLim781** [comment]
also here is a picture of the setup used to test DMA
IMG_3514

File: Project/DAQ_System/Program/Src/Sensor/LinearPotentiometer/ilinear_potentiometer.hpp [hunk 1]
```diff
@@ -20,8 +20,8 @@ class ILinearPotentiometer {
     public:
         virtual ~ILinearPotentiometer() = default;
 
-        virtual float DisplacementInches() = 0;
-        virtual float DisplacementMillimeters() = 0;
+        virtual void DisplacementInches(float data[]) = 0;
+        virtual void DisplacementMillimeters(float data[]) = 0;
```
Comment (crsz20) [inline Project/DAQ_System/Program/Src/Sensor/LinearPotentiometer/ilinear_potentiometer.hpp]: Perhaps name the array within the format of the data being received. I liked the name `quantized_counts` — something more explicit.

**Review by crsz20 — APPROVED** [review]

**Comment by WilliamLim781** [comment]
### further documentation of DMA

the DMA buffer is linked to the ranking of the ADC channels in the .ioc file. This means that the ADC channel that is rank 1 is also the first element of the DMA buffer, rank 2 is the second element, etc. The reason for this is due to the fact that the DMA is set to circular which means that it overwrites the buffer with new values as it gets more values(loops around the buffer). How the DMA overwrites the buffer is based on the ranking of the ADC channels which ensures that each element in the DMA buffer is linked to an ADC channel. 

these are the DMA settings used to achieve the results that I got




dmaSettings


dmaSettings2


there are also things like DMA interrupts that are also part of the setup but these are things that can only change when you change the settings that are in these pictures. Also, I choose 15 clock cycles for each channel for no particular reason. I just saw that when I selected 12-bit resolution it said 15 clock cycles so I chose it. This is something that may need some research later on to see how these sampling times impact performance

File: Project/DAQ_System/Program/Src/Sensor/LinearPotentiometer/ilinear_potentiometer.hpp (hunk 1 above)
Comment (WilliamLim781) [reply Project/DAQ_System/Program/Src/Sensor/LinearPotentiometer/ilinear_potentiometer.hpp]: imma take that variable name from ya

**Review by crsz20 — APPROVED** [review]

**Review by crsz20 — APPROVED** [review]

**Pull request merged**